transaction/std/src/transaction_builder.rs

```diff
@@ -3104,4 +3105,505 @@ pub mod transaction_builder_tests {
+            assert_eq!(funding_amount.token_id, token_id);
+
+            if block_version.e_memo_feature_is_supported() {
+                let funding_change_output_memo = funding_change_output
+                    .e_memo
+                    .unwrap()
+                    .decrypt(&funding_change_output_ss);
+                let funding_output_memo =
+                    funding_output.e_memo.unwrap().decrypt(&funding_output_ss);
+                match MemoType::try_from(&funding_change_output_memo)
+                    .expect("Couldn't decrypt funding change memo")
+                {
+                    MemoType::GiftCodeFunding(memo) => {
+                        assert!(memo.public_key_matches(funding_output_public_key),);
+                        assert_eq!(memo.funding_note().unwrap(), note,);
+                    }
+                    _ => {
+                        panic!("unexpected memo type")
+                    }
+                }
+                match MemoType::try_from(&funding_output_memo)
+                    .expect("Couldn't decrypt gift code funding_output memo")
+                {
+                    MemoType::Unused(_) => {}
+                    _ => {
+                        panic!("unexpected memo type")
+                    }
+                }
+            }
+
+            // MCIP #32 specifies that the receiver will receive the TxOut index,
+            // shared_secret and onetime_private_key in a message. Below we
+            // simulate a receiver sending the gift code to themselves with the
+            // those 3 pieces of information from the gift code TxOut funded above
+
+            // Get the data we're pretending sender sends to the receiver and
+            // construct TxOutGiftCode object from it. This data would
+            // normally be sent via a protobuf message
+            let global_index = 42;
+            let gift_code_tx_out_private_key = recover_onetime_private_key(
+                funding_output_public_key,
+                sender.spend_private_key(),
+                &sender.gift_code_subaddress_spend_private(),
+            );
+            let tx_out_gift_code = TxOutGiftCode {
+                global_index,
+                onetime_private_key: gift_code_tx_out_private_key,
+                shared_secret: funding_output_ss,
+            };
+
+            // Values we pretend we get from locating the TxOut using the global index
+            let masked_amount = funding_output.masked_amount.clone();
+
+            // Construct the sender Tx from the combo of "located" and "sent" information
+            let (sending_input_amount, blinding) = masked_amount
+                .get_value(&tx_out_gift_code.shared_secret)
+                .unwrap();
+            println!("{:?}", sending_input_amount);
+            let sending_output_amount = Amount::new(sending_input_amount.value - fee, token_id);
+
+            let (ring, real_index) = get_ring_for_txout(
```

Comment:
Is it possible to instead use `get_ring` (https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/std/src/test_utils.rs#L65) and replace the TxOut at the `real_index` with `funding_output`? Not sure why the new `get_ring_for_txout` is needed and if we can reduce the amount of test utilities that generate fake data, it will benefit us long term.

Reply:
I built it to ensure the gift code flow where you're spending a TxOut that belongs to another account worked. I could remove the general method and just put the boilerplate into the actual test file.